src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -111,10 +111,23 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
 
         public override bool TryRead(out ReadResult readResult)
         {
-            var result = _context.RequestBodyPipe.Reader.TryRead(out readResult);
-            _readResult = readResult;
+            TryStart();
+
+            var hasResult = _context.RequestBodyPipe.Reader.TryRead(out readResult);
+
+            if (hasResult)
+            {
+                _readResult = readResult;
+
+                StopTimingRead(readResult.Buffer.Length);
```

Comment:
Where is this coming from? We should only call StopTimingRead if we saw an actually-async read and therefore called StartTimingReadAsync.

Reply:
`StartTimingReadAsync` sets `_backpressure = true` which enables the other path in `StopTimingRead`

However it still needs to call `_context.TimeoutControl.BytesRead(bytesRead - _alreadyTimedBytes)` which is called in `ReadAsync` even if it doesn't go async, so it still needs to call that for `TryRead` for that reporting.

Otherwise these three tests fail:
```
DATA_Received_TooSlowlyOnLargeRead_AbortsConnectionAfterRateTimeout
DATA_Received_TooSlowlyOnMultipleStreams_AbortsConnectionAfterAdditiveRateTimeout
DATA_Received_TooSlowlyOnSecondStream_AbortsConnectionAfterNonAdditiveRateTimeout
```
With the falling errors
```
Message: 
    Moq.MockException : 
    Expected invocation on the mock should never have been performed, but was 32 times: h => h.OnTimeout(It.IsAny<TimeoutReason>())
    
    Configured setups: 
    ITimeoutHandler h => h.OnTimeout(It.IsAny<TimeoutReason>())
    
    Performed invocations: 
    ITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)
    ...
    ITimeoutHandler.OnTimeout(TimeoutReason.ReadDataRate)
  Stack Trace: 
    at Mock.VerifyCalls(Mock targetMock, InvocationShape expectation, LambdaExpression expression, Times times, String failMessage)
    at Mock.Verify[T](Mock`1 mock, Expression`1 expression, Times times, String failMessage)
    at Mock`1.Verify(Expression`1 expression, Times times)
    at Mock`1.Verify(Expression`1 expression, Func`1 times)
    at Http2TimeoutTests.DATA_Received_TooSlowlyOnSecondStream_AbortsConnectionAfterNonAdditiveRateTimeout() in Http2TimeoutTests.cs line: 776
    at --- End of stack trace from previous location where exception was thrown ---
```

Reply:
The thing I think we should be calling for an "instant" read is `_context.TimeoutControl.BytesRead(long)`. _context is a private field on MessageBody, so I'd probably just expose it as a protected method. Currently, BytesRead is only called by MessageBody.StopTiming

This is pretty much what already happens when you call MessageBody.StartTimingReadAsync(readAwaitable, token) with an already-completed readAwaitable, and then immediately call MessageBody.StopTimingRead(bytesRead). The MessageBody.TryRead path has never been as well tested as the MessageBody.ReadAsync path, since before now, MessageBody.TryRead was pretty much exclusively used by MessageBody.ConsumeAsync().

Looking at MessageBody, it looks like we can just remove _alreadyTimedBytes. When we refactored the \*MessageBody types, we wound up doing the copy from ReadResult to the Stream.ReadAsync buffer above the MessageBody layer so _alreadyTimedBytes is completely unused. I would almost expect the compiler to catch it, but it doesn't since we "reset" it to 0. If you want to fix that in the PR, that'd be appreciated, otherwise I'll probably clean it up later.

Reply:
So maybe move it out to another method so its more explicit? Will do.

Reply:
> We should only call StopTimingRead if we saw an actually-async read and therefore called StartTimingReadAsync

I was a wrong about this as you noted. What I should have said is we should only call StopTimingRead if we called StartTimingReadAsync. It doesn't matter whether or not it was actually async.

Reply:
Made change